README.md
koa-better-ratelimit [![Build Status](https://travis-ci.org/tunnckoCore/koa-better-ratelimit.png)](https://travis-ci.org/tunnckoCore/koa-better-ratelimit) [![Dependency Status](https://david-dm.org/tunnckoCore/koa-better-ratelimit.svg)](https://david-dm.org/tunnckoCore/koa-better-ratelimit)
================

> Better, smaller, faster. KoaJS middleware for limit request by ip, store in-memory.

## Install
> Install [`package/koa-better-ratelimit`](http://npm.im/koa-better-ratelimit) with [npm](https://npmjs.org)

```
$ npm install koa-better-ratelimit
```

## Options
> With options through init you can control black/white lists, limit per ip and reset interval.

- `duration` **{Integer}** Limit duration in seconds, default `1000 * 60 * 60 * 1` (1 hour)
- `whiteList` **{Array}** All ips that won't be limited, default `empty array`
- `blackList` **{Array}** All ips that always be limited and 403, default `empty array`
- `message_429` **{String}** Message for all requests after limit, default `429: Too Many Requests.`
- `message_403` **{String}** Message for limited/forbidden, default `403: This is forbidden area for you.`
- `max` **{Integer}** Max requests per ip, default `500`
- `env` **{Boolean}** Manage enviroment, for tests will use `x-koaip` header, default `null`


## Usage
> Some demo example which is exactly `example.js`

```js
var app   = require('koa')(),
    limit = require('./index');

app.use(limit({
  duration: 1000 * 60 * 3, //3 mins
  max: 5,
  //blackList: ['127.0.0.1']
}));

app.use(function * (next) {
  this.body = 'Hello World';
  yield next;
});

var port = process.env.PORT || 3333;
app.listen(3333);
console.log('Koa server start listening on port ' + port);
```

## Diferences
> Between [koa-better-ratelimit](https://github.com/tunnckoCore/koa-better-ratelimit) and [koa-ratelimit](https://github.com/koajs/ratelimit)

- Support blackList and whiteList options
- Pure in-memory store, no other adapters
- `duration` option in seconds, not in milliseconds
- 7 working tests
- removed `db` option
- added `Retry-After` header
- added separate 403 and 429 option messages

> Between [koa-better-ratelimit](https://github.com/tunnckoCore/koa-better-ratelimit) and [koa-limit](https://github.com/koajs/koa-limit)

- koa-limit [is totally broken](https://github.com/koajs/koa-limit/issues/3#issuecomment-42731409) (to v1.0.1)
- removed `redis` and test dependencies
- smaller, better, working, simple
- added separate 403 and 429 option messages


## Test, Bench, Example
> First run `npm install`

```
$ npm test
$ npm start
```


## License
The MIT License, 2014 [Charlike Mike Reagent](https://github.com/tunnckoCore) ([@tunnckoCore](https://twitter.com/tunnckoCore))
***
_This file was generated by [`Ock`](https://github.com/tosckjs/ock) on June 20, 2014._


history.md
## v2.1.2 / November 27, 2014
- bump to `jscs >= 1.8.0` and `jscs-doc >= 0.2.0`
- update semver range

## v2.1.1 / November 27, 2014
- fix code style collisions

## v2.1.0 / November 27, 2014
- enforce new code style
  - add jshint, jshintignore, jscs (using jscs-jsdoc)
  - normalizing all @tunnckoCore packages
  - add/edit dotfiles 
- option `message_429` deprecated, it will be remove in **v2.2.x**, use `accessLimited`
- option `message_403` deprecated, it will be remove in **v2.2.x**, use `accessForbidden`
- _>=v2.2.x jscs: "requireCamelCaseOrUpperCaseIdentifiers: true"_
- _>=v2.2.x jshint: "camelcase: true"_

## v2.0.1 / October 27, 2014
- typo in readme, `options.duration` is in milliseconds

## v2.0.0 / October 27, 2014
- remove `copy-to` dependency
- full rewrite
- fix wrong behaving - now works correctly
- MAY break someone's behaving

## v1.0.1 / October 25, 2014
- fix typos
- fix package.json

## v1.0.0 / October 25, 2014
- initial release



[npmjs-url]: http://npm.im/koa-better-ratelimit
[npmjs-shields]: http://img.shields.io/npm/v/koa-better-ratelimit.svg
[npmjs-install]: https://nodei.co/npm/koa-better-ratelimit.svg?mini=true

[coveralls-url]: https://coveralls.io/r/tunnckoCore/koa-better-ratelimit?branch=master
[coveralls-shields]: https://img.shields.io/coveralls/tunnckoCore/koa-better-ratelimit.svg

[license-url]: https://github.com/tunnckoCore/koa-better-ratelimit/blob/master/license.md
[license-img]: http://img.shields.io/badge/license-MIT-blue.svg

[travis-url]: https://travis-ci.org/tunnckoCore/koa-better-ratelimit
[travis-img]: https://travis-ci.org/tunnckoCore/koa-better-ratelimit.svg?branch=master

[depstat-url]: https://david-dm.org/tunnckoCore/koa-better-ratelimit
[depstat-img]: https://david-dm.org/tunnckoCore/koa-better-ratelimit.svg

[author-gittip-img]: http://img.shields.io/gittip/tunnckoCore.svg
[author-gittip]: https://www.gittip.com/tunnckoCore
[author-github]: https://github.com/tunnckoCore
[author-twitter]: https://twitter.com/tunnckoCore

[author-website]: http://www.whistle-bg.tk
[author-npmjs]: https://npmjs.org/~tunnckocore

[koa-url]: https://github.com/koajs/koa

test.js
/**
 * koa-better-ratelimit <https://github.com/tunnckoCore/koa-better-ratelimit>
 *
 * Copyright (c) 2014 Charlike Mike Reagent, contributors.
 * Released under the MIT license.
 */

'use strict';

/**
 * Module dependencies.
 */

var koa = require('koa');
var limit = require('./index');
var request = require('supertest');
var http = require('http');

function helloMiddleware() {
  return function * hello(next) {
    this.response.type = 'text/html'
    this.response.body = '<p>Hello test.</p>';
    yield * next;
  }
}

/**
 * appNonDefault description
 */
var appNonDefault = koa();
appNonDefault.use(limit({
  duration: 1000 * 60, //1 min
  max: 3, //max requests
  env: 'test'
}));
appNonDefault.use(helloMiddleware());
appNonDefault = http.createServer(appNonDefault.callback());

/**
 * appBlack description
 */
var appBlack = koa();
appBlack.use(limit({
  blackList: ['4.4.1.*'],
  accessForbidden: 'access forbidden, please contact foo@bar.com',
  env: 'test'
}));
appBlack.use(helloMiddleware());
appBlack = http.createServer(appBlack.callback());

/**
 * appWhite description
 */
var appWhite = koa();
appWhite.use(limit({
  whiteList: ['127.0.*.*'],
  max: 50,
  env: 'test'
}));
appWhite.use(helloMiddleware());
appWhite = http.createServer(appWhite.callback());

describe('appNonDefault', function() {
  it('should status 200 - 1.2.3.4 - remaining 2/3', function(done) {
    request(appNonDefault)
    .get('/')
    .set('x-koaip', '1.2.3.4')
    .expect(200, '<p>Hello test.</p>')
    .expect('X-RateLimit-Limit', '3')
    .expect('X-RateLimit-Remaining', '2')
    .end(done);
  });
  it('should status 200 - 1.2.3.4 - remaining 1/3', function(done) {
    request(appNonDefault)
    .get('/')
    .set('x-koaip', '1.2.3.4')
    .expect(200, '<p>Hello test.</p>')
    .expect('X-RateLimit-Limit', '3')
    .expect('X-RateLimit-Remaining', '1')
    .end(done);
  });
  it('should status 200 - 1.2.3.4 - remaining 0/3', function(done) {
    request(appNonDefault)
    .get('/')
    .set('x-koaip', '1.2.3.4')
    .expect(200, '<p>Hello test.</p>')
    .expect('X-RateLimit-Limit', '3')
    .expect('X-RateLimit-Remaining', '0')
    .end(done);
  });
  it('should status 429 - 1.2.3.4 - remaining 0/3 /1', function(done) {
    request(appNonDefault)
    .get('/')
    .set('x-koaip', '1.2.3.4')
    .expect(429, '429: Too Many Requests.')
    .expect('X-RateLimit-Limit', '3')
    .expect('X-RateLimit-Remaining', '0')
    .end(done);
  });
  it('should status 200 - 8.8.8.8 - remaining 2/3', function(done) {
    request(appNonDefault)
    .get('/')
    .set('x-koaip', '8.8.8.8')
    .expect(200, '<p>Hello test.</p>')
    .expect('X-RateLimit-Limit', '3')
    .expect('X-RateLimit-Remaining', '2')
    .end(done);
  });
  it('should status 200 - 8.8.8.8 - remaining 1/3', function(done) {
    request(appNonDefault)
    .get('/')
    .set('x-koaip', '8.8.8.8')
    .expect(200, '<p>Hello test.</p>')
    .expect('X-RateLimit-Limit', '3')
    .expect('X-RateLimit-Remaining', '1')
    .end(done);
  });
  it('should status 429 - 1.2.3.4 - remaining 0/3 /2', function(done) {
    request(appNonDefault)
    .get('/')
    .set('x-koaip', '1.2.3.4')
    .expect(429, '429: Too Many Requests.')
    .expect('X-RateLimit-Limit', '3')
    .expect('X-RateLimit-Remaining', '0')
    .end(done);
  });
  it('should status 200 - 8.8.8.8 - remaining 0/3', function(done) {
    request(appNonDefault)
    .get('/')
    .set('x-koaip', '8.8.8.8')
    .expect(200, '<p>Hello test.</p>')
    .expect('X-RateLimit-Limit', '3')
    .expect('X-RateLimit-Remaining', '0')
    .end(done);
  });
  it('should status 429 - 8.8.8.8 - remaining 0/3', function(done) {
    request(appNonDefault)
    .get('/')
    .set('x-koaip', '8.8.8.8')
    .expect(429, '429: Too Many Requests.')
    .expect('X-RateLimit-Limit', '3')
    .expect('X-RateLimit-Remaining', '0')
    .end(done);
  });
});

describe('appBlackList', function() {
  it('should 403 Forbidden - blackList', function(done) {
    request(appBlack)
    .get('/')
    .set('x-koaip', '4.4.1.8')
    .expect(403, 'access forbidden, please contact foo@bar.com')
    .end(done);
  });
});

describe('appWhiteList', function() {
  it('should 200 OK - whiteList - no limits', function(done) {
    request(appWhite)
    .get('/')
    .set('x-koaip', '127.0.4.4')
    .expect(200, '<p>Hello test.</p>')
    .end(done);
  });
});


index.js
/**
 * koa-better-ratelimit <https://github.com/tunnckoCore/koa-better-ratelimit>
 *
 * Copyright (c) 2014 Charlike Mike Reagent, contributors.
 * Released under the MIT license.
 */

'use strict';

/**
 * Module dependencies.
 */

var ipchecker = require('ipchecker');

var defaults = {
  duration: 1000 * 60 * 60 * 24,
  whiteList: [],
  blackList: [],
  accessLimited: '429: Too Many Requests.',
  accessForbidden: '403: This is forbidden area for you.',
  max: 500,
  env: null
};

/**
 * With options through init you can control
 * black/white lists, limit per ip and reset interval.
 *
 * @param {Object} options
 * @api public
 */
module.exports = function betterlimit(options) {
  options = options || {};

  var db = {};

  for (var key in defaults) {
    if (!options[key]) {options[key] = defaults[key]}
  }

  if (options.message_429) {
    options.accessLimited = options.message_429;
  }

  if (options.message_403) {
    options.accessForbidden = options.message_403;
  }

  var whiteListMap = ipchecker.map(options.whiteList);
  var blackListMap = ipchecker.map(options.blackList);

  return function * ratelimit(next) {
    var ip = options.env === 'test' ? this.request.header['x-koaip'] : this.ip;

    if (!ip) {
      return yield * next;
    }
    if (ipchecker.check(ip, blackListMap)) {
      this.response.status = 403;
      this.response.body = options.accessForbidden;
      return;
    }
    if (ipchecker.check(ip, whiteListMap)) {
      return yield * next;
    }

    var now = Date.now()
    var reset = now + options.duration;

    if (!db.hasOwnProperty(ip)) {
      db[ip] = {ip: ip, reset: reset, limit: options.max}
    }

    var delta = db[ip].reset - now
    var retryAfter = delta / 1000 | 0;

    db[ip].limit = db[ip].limit - 1
    this.response.set('X-RateLimit-Limit', options.max);

    if (db[ip].reset > now) {
      var rateLimiting = db[ip].limit < 0 ? 0 : db[ip].limit;
      this.response.set('X-RateLimit-Remaining', rateLimiting);
    }

    if (db[ip].limit < 0 && db[ip].reset < now) {
      db[ip] = {ip: ip, reset: reset, limit: options.max}
      db[ip].limit = db[ip].limit - 1;
      this.response.set('X-RateLimit-Remaining', db[ip].limit);
    }

    this.response.set('X-RateLimit-Reset', db[ip].reset);

    if (db[ip].limit < 0) {
      this.response.set('Retry-After', retryAfter);
      this.response.status = 429;
      this.response.body = options.accessLimited
      return;
    }

    return yield * next
  };
}
